.pre-commit-config.yaml

```diff
@@ -0,0 +1,24 @@
+# Automatically run checks over changed files prior to `git commit`.
+# Required one-time setup:
+#
+#   pre-commit install
+#
+# See https://pre-commit.com for details.
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.4.0
+    hooks:
+      - id: trailing-whitespace
+      - id: end-of-file-fixer
+
+  - repo: https://github.com/psf/black
+    rev: 23.3.0
+    hooks:
+      - id: black
+        args: ["."]
+
+  - repo: https://github.com/pycqa/isort
+    rev: 5.12.0
+    hooks:
+      - id: isort
+        args: ["."]
```

Comment:
> is it a good idea to have two slightly different sets of linting passes (tox -e lint and this pre-commit list)?

I don't see a big problem there. Config is the same for both. Differences would mainly come from different versions of black && isort. But `tox -e lint` on CI is the ultimate authority. 

> do we understand the dependency chains here? I assume black and isort hooks depend on already installed tools (or does pre-commit install magically install them from somewhere?)... but what about the "builtin" hooks? Do they use code embedded in the pre-commit package?

I admit, I don't exactly know where pre-commit installs the dependencies, but I suppose in some virtual environment. It [claims](https://pre-commit.com/#python) that *"python hooks work without any system-level dependencies"*.

> are we in trouble when we inevitably forget to update these versions? I suppose not?

I don't think, we are. But yeah, it would be nice to automate updates. There are some 3rd-party solutions to do this, and a feature request for Dependabot.